### ISSUE TITLE
YARA UI & Service Updates

### ISSUE THREAD

**Issue by eljeffeg** [CybercentreCanada/assemblyline#55]
1. It would be nice if the Signature Search also searched on `Source`.  So if I search bartblaze or reversinglabs, I'd get some Yara hits for Signatures.  I couldn't find the Yara (perhaps due to a 10,000 limit on Signatures) so I was searching on the Source and kept getting 0 hits.  I didn't realize I needed to search Yara.  I did end up finding them by clicking on the source fingerprint.

2. Some of my yara source updates don't appear to be processing the yara rules.  Here are couple public repositories that I've added that say they've downloaded, but show no signatures.

    - (*.yara$) https://github.com/chronicle/GCTI/tree/main/YARA
    - (*.yara$) https://github.com/reversinglabs/reversinglabs-yara-rules.git

**Comment by cccs-jp** [comment]
for number 1 you can do: source:reversinglabs  (case sensitive)
@cccs-sgaron Will investigate indexing all the text of the rule body in the futur.
@cccs-rs for #2 :)

**Comment by cccs-rs** [comment]
Regex seems wrong, should be `.*\.yara$`

**Comment by cccs-rs** [comment]
Refer to my recent messages re: 1. https://discord.com/channels/908084610158714900/908084610624274494/1086247172724510843

**Comment by cccs-rs** [comment]
Source Configurations (typo on the second but still matched the file path):
![image](https://user-images.githubusercontent.com/62077998/225895773-8ae667be-ded9-49f9-acd6-c15878ac4f6d.png)

Which yielded signatures being import for both sources:
![image](https://user-images.githubusercontent.com/62077998/225895467-2531a428-9015-475d-a75a-809548a68114.png)


Is there anything in the `yara-updates` container to suggest what the issue could be?

**Comment by eljeffeg** [comment]
No errors in the logs.  For all purposes that I can tell, it completes successfully. There are just no signatures when I click on the fingerprint.  `Status: Skipped.`
![Screenshot 2023-03-17 at 9 41 41 AM](https://user-images.githubusercontent.com/161904/225921647-aec870e2-abef-4311-95af-061a4b14b754.png)
![Screenshot 2023-03-17 at 9 41 32 AM](https://user-images.githubusercontent.com/161904/225921650-d7f3aa52-cc0f-4d18-b923-ae1411542893.png)
![Screenshot 2023-03-17 at 9 41 48 AM](https://user-images.githubusercontent.com/161904/225921645-b977e9a7-a6ee-4a86-9052-f6f54a8ab9e4.png)

**Comment by cccs-rs** [comment]
Issue partially resolved by clearing the cached entries in Redis. Adding a cache-less update mechanism has been created as a ticket.

**Comment by eljeffeg** [comment]
Documenting what @cccs-rs and I have discussed in chat (and also a reminder).  The Redis cache resolved 2 of my 3 issues with the yara files.  The last was reported an error of `'utf-8' codec can't decode byte 0xbb in position 908: invalid start byte.`

The line causing the exception is here:  https://github.com/CybercentreCanada/assemblyline-service-yara/blob/3061ac084004db4cb514d192ab88fd142ad9d09e/yara_/update_server.py#L96

The `readlines()` causes an exception as it finds a character it can't decode.  I did verify the files are utf-8 files that just have unknown characters in them.  [Here is an example](https://github.com/mattulm/sfiles_yara/blob/ddeb9cf0232cd546e9f1b31df4539c6c0db32832/malware/Inception.yar).

What I guess I'd suggest is maybe a couple things... 

1. Might make sense to Include the reading of the file in the try / except statement, so even if there is an error on a file, it continues to read the other rules.

2. Probably would be good to define errors in the `open(errors=)` so that it can hopefully use the rule as intended.  I'm not sure if they should be set to `ignore` or maybe `surrogateescape`? https://docs.python.org/3/library/functions.html#open

**Comment by malvidin** [comment]
Based on these comments, I recommend ignoring any encoding errors.
https://github.com/VirusTotal/yara/pull/1770#issuecomment-1357622486
https://github.com/VirusTotal/yara-python/pull/136#issuecomment-612820721

**Comment by malvidin** [comment]
> $str_08 = "<(:<\\Documents and Settings\\all users\\Application Data\\�"

Do you have a link to the source, or can you upload that portion of the rule here? Strings pasted inline won't retain the source encoding.

**Comment by eljeffeg** [comment]
It's not public, but here is a modified file based on that rule.

**Comment by eljeffeg** [comment]
It's not public, but here is a modified file based on that rule.
[testgen.txt](https://github.com/CybercentreCanada/assemblyline/files/11123398/testgen.txt)

**Comment by eljeffeg** [comment]
Yeah, it was a `BB` hex for that one, which comes across in my hex editor as `»`.

**Comment by malvidin** [comment]
Looks like I was behind on the status of the rule encodings. But there appears to be a consistent desire from the development team to only support UTF-8 in the future.
https://github.com/VirusTotal/yara/commit/68eb237dd66dd5805cb44ce17a3b22c999f43bbe

`iso-8859-1` will accept any byte, but the values in `[\x80-\xFF]` may be incorrect if a different encoding was used at the source.
```python
                try:
                    with open(file, 'r') as f:
                        f_lines = f.readlines()
                except UnicodeDecodeError as exc:
                    with open(file, 'r', encoding='iso-8859-1') as f:
                        f_lines = f.readlines()
                    self.log.error(f"File could not be loaded as UTF-8: {file}")
```

**Comment by cccs-rs** [comment]
[v4.4.0.stable5](https://github.com/CybercentreCanada/assemblyline-service-yara/releases/tag/v4.4.0.stable5) of the YARA service will include performing a `surrogateescape` when reading the contents of YARA files with non-UTF8 charactors

Services w/ updaters built from [v4.4.0.stable5](https://github.com/CybercentreCanada/assemblyline/releases/tag/v4.4.0.stable5) of Assemblyline will now be able to instruct the Scaler about when to scale a service. 

If a service depends on the bundle sent by the updater, indicated by `wait_for_update` flag, (ie. YARA) then the service is only considered 'active' and ready for scaling only if the updater deems it so (ie. it has at least 1 downloaded source).

If you notice this issue still persists, feel free to re-open the issue with more information! 😁
